docs/rfcs/confmap-provider-configuration.md

```diff
@@ -0,0 +1,184 @@
+tasked with retrieving from config from all URIs passed to the Collector for the
+scheme the provider is registered to handle.
+
+We have the following places where it may make sense to hook into options for
+Providers:
+
+1. Parts of the URI we are requesting.
+2. Part of the string passed to `--config`.
+3. Separate flags to configure Providers per config URI.
+4. Use a separate config file that specifies config sources inside a map
+   structure.
+5. Extend the Collector's config schema to support specifying additional places
+   to obtain configuration
+
+### Configure options inside the URI
+
+[RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986#section-3), which
+specifies the format of a URI, specifies the different parts of a URI and
+suggests two places where we could pass options to Providers: queries and
+fragments.
+
+#### Queries
+
+Advantages:
+
+- Explicitly intended to  specify non-hierarchical data in a URI.
+- Often used for this purpose.
+- Fits into existing config URIs.
+
+Disadvantages:
+
+- Completely consuming the parameters would mean we can't pass them to config
+  backends. The frequent use of query parameters in server endpoints would force
+  us to limit the types of requests we can make.
+- Partially consuming parameters would still cause conflicts if a Provider
+  parameter and config backend parameter share the same name. Additionally,
+  mixing the two would provide suboptimal UX.
+- Only allows easily specifying key-value pairs.
+
+#### Fragments
+
+We could specify options in a query parameter-encoded string placed into the URI
+fragment.
+
+Advantages:
+
+- Not likely to be used by config backends for any of our supported protocols,
+  so has a low chance of conflict.
+- Fits into existing config URIs.
+
+Disadvantages:
+
+- Even if fragments are likely not useful to backends, we are still preventing
+  their use in upstream Providers.
+- Doesn't conform to the spirit of how fragments should be used according to RFC
+  3986.
+- Only allows easily specifying key-value pairs.
+
+We could likely partially circumvent the key-value pair limitation by
+recursively calling confmap Providers to resolve files, env vars, HTTP URLs,
+etc. For example:
+
+```text
+https://config.com/config#refresh-interval=env:REFRESH_INTERVAL&headers=file:headers.yaml
+```
+
+Using this strategy would also allow us to more easily get env vars and to get
+values from files for things like API tokens.
+
+### Add to the string passed to the config flag
+
+Instead of configuring Providers inside their URIs, we could add an extra
+section to the string passed to `--config` and configure how a URI is retrieved
+through that config. This approach is undesirable because the config flag
+currently only takes valid URIs, and updating the format to accommodate this
+would require we adopt an unconventional format.
+
+### Separate flags to configure Providers per config URI
```

Comment:
I am not entirely opposed to allow configuring things via flags, but if we ever do this I would do that with a `flag` provider that  you can call like `${flag:my-flag-name}` and it would give you the (YAML-parsed?) value of `--my-flag-name=val` (or possibly `--some-namespace-here-my-flag-name`).

This would have to piggyback onto some other solution that allows passing this `${flag:my-flag-name}` value though

Reply:
If we were to take this approach, how could we determine which flags are valid? It sounds like flag names could be arbitrary, so would passing `--my-unsupported-flag` cause an error, or just store a value in the flag provider that goes unused? Or am I entirely misunderstanding your suggestion?

Reply:
> If we were to take this approach, how could we determine which flags are valid? It sounds like flag names could be arbitrary, so would passing --my-unsupported-flag cause an error, or just store a value in the flag provider that goes unused? Or am I entirely misunderstanding your suggestion?

I was thinking about requiring the flags to be under some namespace (e.g. `--flag-provider-my-flag-name`). I am not sure what the behavior should be for unused flags under that namespace (ideally an error at config resolution time if they go unused).

Reply:
Thanks for explaining.

The way I would see this working right now is to make config flags order-sensitive, like so:

```
--config=http://myurl.com
--flag-provider-refresh-interval=10s
--flag-provider-auth-header=${env:MYPASS}
--config=/path/to/file.yaml
[...]
```

We would then need to give the Provider's `Retrieve` method some kind of handle to obtain these flag values from. Do you think this would be viable, or do you have another solution you think may be better here?